File: lib/src/m3u8_source_type.dart
enum M3u8SourceType {
  auto,
  hls,
  progressive;

  String get platformValue => switch (this) {
    M3u8SourceType.auto => 'auto',
    M3u8SourceType.hls => 'hls',
    M3u8SourceType.progressive => 'progressive',
  };

  static M3u8SourceType from(String? value) {
    return switch (value) {
      'hls' => M3u8SourceType.hls,
      'progressive' => M3u8SourceType.progressive,
      _ => M3u8SourceType.auto,
    };
  }
}

File: android/src/main/kotlin/com/ai3/player_m3u8/player_m3u8/M3u8SourceType.kt
package com.ai3.player_m3u8.player_m3u8

enum class M3u8SourceType {
    AUTO,
    HLS,
    PROGRESSIVE;

    fun resolve(url: String): M3u8SourceType {
        if (this != AUTO) {
            return this
        }
        val path = url.substringBefore('?').substringBefore('#').lowercase()
        return when {
            path.endsWith(".m3u8") -> HLS
            path.endsWith(".mp4") || path.endsWith(".mov") -> PROGRESSIVE
            else -> HLS
        }
    }

    fun platformValue(): String {
        return when (this) {
            AUTO -> "auto"
            HLS -> "hls"
            PROGRESSIVE -> "progressive"
        }
    }

    companion object {
        fun from(value: String?): M3u8SourceType {
            return when (value?.lowercase()) {
                "hls" -> HLS
                "progressive" -> PROGRESSIVE
                else -> AUTO
            }
        }
    }
}
